<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Open-and-examine-the-file" data-toc-modified-id="Open-and-examine-the-file-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Open and examine the file</a></span><ul class="toc-item"><li><span><a href="#We-import-the-modules-and-libraries-necessary-for-work" data-toc-modified-id="We-import-the-modules-and-libraries-necessary-for-work-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>We import the modules and libraries necessary for work</a></span></li><li><span><a href="#Opening-the-table" data-toc-modified-id="Opening-the-table-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Opening the table</a></span></li><li><span><a href="#Let's-review-the-basic-data-about-the-dataset" data-toc-modified-id="Let's-review-the-basic-data-about-the-dataset-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Let's review the basic data about the dataset</a></span></li></ul></li><li><span><a href="#Divide-the-data-into-samples" data-toc-modified-id="Divide-the-data-into-samples-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Divide the data into samples</a></span><ul class="toc-item"><li><span><a href="#Let's-denote-the-features-as-features,-and-the-target-feature-as-target" data-toc-modified-id="Let's-denote-the-features-as-features,-and-the-target-feature-as-target-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Let's denote the features as <code>features</code>, and the target feature as <code>target</code></a></span></li><li><span><a href="#We-divide-the-initial-data-into-training,-validation-and-test-sets.-Let's-take-the-ratio-of-60/20/20%,-respectively" data-toc-modified-id="We-divide-the-initial-data-into-training,-validation-and-test-sets.-Let's-take-the-ratio-of-60/20/20%,-respectively-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>We divide the initial data into training, validation and test sets. Let's take the ratio of 60/20/20%, respectively</a></span></li></ul></li><li><span><a href="#Exploring-Models" data-toc-modified-id="Exploring-Models-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Exploring Models</a></span><ul class="toc-item"><li><span><a href="#Using-a-decision-tree" data-toc-modified-id="Using-a-decision-tree-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Using a decision tree</a></span></li><li><span><a href="#Let's-try-random-forest" data-toc-modified-id="Let's-try-random-forest-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Let's try random forest</a></span></li><li><span><a href="#Now-let's-check-the-logistic-regression" data-toc-modified-id="Now-let's-check-the-logistic-regression-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Now let's check the logistic regression</a></span></li></ul></li><li><span><a href="#Let's-check-the-model-on-the-test-set" data-toc-modified-id="Let's-check-the-model-on-the-test-set-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Let's check the model on the test set</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

# Tariff recommendation

The mobile operator "Megaline" found out that many customers use archival tariffs. They want to build a system that can analyze customer behavior and offer users a new tariff: "Smart" or "Ultra".

We are faced with the task of classification.



We have data on the behavior of customers who have already switched to these tariffs (from the draft course "Statistical Data Analysis"). You need to build a model for the classification problem that will select the appropriate rate. Data preprocessing is not required - we have already done it.

Let's build a model with the maximum *accuracy* value. Let's check *accuracy* on the test set.

Let's open the data file and examine it.

We divide the initial data into training, validation and test sets.

We investigate the quality of different models by changing the hyperparameters. Let us briefly write the conclusions of the study.

Let's check the quality of the model on a test sample.


Data Description

Each object in the data set is information about the behavior of one user per month. Known:

calls — number of calls,

minutes — total duration of calls in minutes,

messages — number of sms messages,

mb_used - Internet traffic used in Mb,

is_ultra - what tariff did you use during the month ("Ultra" - 1, "Smart" - 0).

## Open and examine the file

### We import the modules and libraries necessary for work

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from joblib import dump
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score

### Opening the table

In [2]:
try:
    data = pd.read_csv('/datasets/users_behavior.csv', sep=',')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/users_behavior.csv', sep=',')

In [3]:
data.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

Each object in the data set is information about the behavior of one user per month. Known:

calls — number of calls,

minutes — total duration of calls in minutes,

messages — number of sms messages,

mb_used - Internet traffic used in Mb,

is_ultra - what tariff did you use during the month ("Ultra" - 1, "Smart" - 0).

### Let's review the basic data about the dataset

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
data.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [6]:
data.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [7]:
data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [8]:
#data['is_ultra'].value_counts()
for col in data.columns:
    print(data[col].value_counts())
    print('Average value:', data[col].mean())
    print()

69.0     60
68.0     49
64.0     48
65.0     47
77.0     47
         ..
157.0     1
203.0     1
148.0     1
167.0     1
168.0     1
Name: calls, Length: 184, dtype: int64
Average value: 63.03889234598631

0.00      40
564.01     3
2.00       3
1.00       3
270.06     2
          ..
498.91     1
258.82     1
296.78     1
336.91     1
566.09     1
Name: minutes, Length: 3144, dtype: int64
Average value: 438.20878655880523

0.0      497
20.0      52
6.0       46
16.0      45
17.0      44
        ... 
180.0      1
169.0      1
185.0      1
201.0      1
170.0      1
Name: messages, Length: 180, dtype: int64
Average value: 38.28126944617299

0.00        11
0.01         2
19915.42     1
20543.85     1
21522.45     1
            ..
20736.28     1
33988.90     1
10328.70     1
22064.65     1
29480.52     1
Name: mb_used, Length: 3203, dtype: int64
Average value: 17207.673836341008

0    2229
1     985
Name: is_ultra, dtype: int64
Average value: 0.30647168637212197



As you can see, the dataset is good, there are no null values, no duplicates. 40 people have the number of minutes equal to zero, 497 people have the number of messages, 11 people have the number of used megabytes. 30% of people used the "Ultra" tariff, the rest did not.

## Divide the data into samples

### Let's denote the features as `features`, and the target feature as `target`

In [9]:
features = data.drop(['is_ultra'], axis=1)
target = data['is_ultra']

### We divide the initial data into training, validation and test sets. Let's take the ratio of 60/20/20%, respectively

In [10]:
features_train, features_valid, target_train, target_valid = (
    train_test_split(
        features, target, test_size=0.4,
        random_state=12345, stratify = target))

We got a training sample = 60%, a validation sample - 40%. Now let's split the validation in half, and get the validation and test

In [11]:
features_valid, features_test, target_valid, target_test =(
    train_test_split(
        features_valid, target_valid, test_size=0.5,
        random_state=12345, stratify=target_valid))

In [12]:
for sample in [features_train, features_valid, features_test, target_train, target_valid, target_test]:
    print(f'Sample size: {sample.shape}')

Sample size: (1928, 4)
Sample size: (643, 4)
Sample size: (643, 4)
Sample size: (1928,)
Sample size: (643,)
Sample size: (643,)


In [13]:
features_train.shape

(1928, 4)

In [14]:
features_valid.shape

(643, 4)

In [15]:
features_test.shape

(643, 4)

In [16]:
target_train.shape

(1928,)

In [17]:
target_valid.shape

(643,)

In [18]:
target_test.shape

(643,)

## Exploring Models

We have a classification task, so only these models are suitable: decision tree, random forest and logical regression

Train the model on the training set, evaluate its quality and select hyperparameters on the validation set

### Using a decision tree

In [19]:
best_model = None
best_result = 0
best_depth = 0
for depth in range(1, 6):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth)
    model.fit(features_train, target_train) # обучим модель на тренировочной выборке
    predictions_valid = model.predict(features_valid) # получим предсказания модели на валидационной выборке
    result = accuracy_score(target_valid, predictions_valid) # посчитаем значение метрики accuracy на валидационной выборке
    if result > best_result:
        best_model = model
        best_result = result
        best_depth = depth

print("Accuracy of the best model on the validation set:", best_result, "Tree depth:", best_depth)

Accuracy of the best model on the validation set: 0.7853810264385692 Tree depth: 5


Got accuracy 0.78, not bad

### Let's try random forest

In the case of using `RandomForestClassifier`, when iterating over hyperparameters, we will iterate over `n_estimators` and `max_depth`

In [20]:
best_model = None
best_result = 0
best_est = 0
best_depth = 0
for est in range(10, 51, 10):
    for depth in range (1, 11):
        model = RandomForestClassifier(
            random_state=12345, n_estimators=est,
            max_depth=depth, criterion='gini') # initialize the model RandomForestRegressor с параметрами random_state=12345, n_estimators=est и max_depth=depth
        model.fit(features_train, target_train) # train the model on the training set
        predictions_valid = model.predict(features_valid) # get the predictions of the model on the validation выборке
        result = accuracy_score(target_valid, predictions_valid) # calculate the value of the accuracy metric на валидационной выборке
        if result > best_result:
            best_model = model
            best_result = result
            best_est = est
            best_depth = depth

print("Accuracy of the best model on the validation set:",
      best_result, "Number of trees:", best_est, "Max Depth:", best_depth)

Accuracy of the best model on the validation set: 0.8211508553654744 Number of trees: 40 Max Depth: 9


We got an accuracy of 0.82, even better than in the case of the decision tree

### Now let's check the logistic regression

In [21]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train, target_train) #train the model
dump(model, 'model_9_1.joblib')
predictions_valid = model.predict(features_valid) # get model predictions on the validation set
accuracy = accuracy_score(target_valid, predictions_valid)
accuracy

0.7387247278382582

We got an accuracy value of 0.74, less than necessary.

The best results are for the Random Forest model, and we will consider it further

## Let's check the model on the test set

We select the best model on the validation sample, we have a random forest

Therefore, now on the test sample we will evaluate this particular model

In [22]:
best_model = None
best_result = 0
for est in range(10, 51, 10):
    for depth in range (1, 11):
        model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = depth) # обучим модель с заданным количеством деревьев
        model.fit(features_train, target_train) # train the model on the training set
        predictions_test = model.predict(features_test) #make predictions
        result = accuracy_score(target_test, predictions_test) # calculate the quality of the model on the test выборке
        if result > best_result:
            best_model = model # save the best model
            best_result = result # save the best value of the accuracy metric on the test data
            best_est = est # keep the depth
            best_depth = depth # keep the number of trees

print("Accuracy of the best model on the test set:",
      best_result, "Number of trees:", best_est, "Depth:", best_depth)

Accuracy of the best model on the test set: 0.8211508553654744 Number of trees: 20 Depth: 6


We got a value of 0.82, better than the required minimum of 0.75

## Conclusion

The purpose of the study was to analyze the behavior of customers and create a system of recommendations for a certain tariff - "Smart" or "Ultra".

The classification task was performed. We have built a model that will select the appropriate tariff with the best value of the *accuracy* metric

The following models were considered:

Decision tree, got the value *accuracy* = 0.78 on the validation set

Random forest, got *accuracy* = 0.8211508553654744 on the validation set with number of trees = 40, depth = 9

Logistic regression, got the value *accuracy* = 0.74 on the validation set


**Having predicted the "Random Forest" model on the test sample, we obtained the metric value accuracy = 0.8211508553654744 with hyperparameters n_estimators (number of trees) = 20, depth (depth) = 6**